File: apps/android/app/src/main/java/pro/wewed/app/ui/roles/RoleShellScaffold.kt
package pro.wewed.app.ui.roles

import androidx.compose.foundation.background
import androidx.compose.foundation.horizontalScroll
import androidx.compose.foundation.layout.BoxWithConstraints
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.*
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.ExperimentalComposeUiApi
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.semantics.testTagsAsResourceId
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import pro.wewed.app.models.NativeDeepLink
import pro.wewed.app.navigation.DeepLinkRouter
import pro.wewed.app.navigation.Entitlements
import pro.wewed.app.navigation.IANavigationContract
import pro.wewed.app.navigation.NavigationContext
import pro.wewed.app.navigation.PrimaryDestination
import pro.wewed.app.navigation.RoleShellAuthorization
import pro.wewed.app.theme.WeddingIdentityPalette

/**
 * Contract-driven Level-1 shell shared by every IA V2 role.
 *
 * Bottom navigation is rendered from [IANavigationContract] rather than hand-written per role, so
 * labels and order cannot drift between roles or platforms. Every destination switch resolves
 * through [Entitlements.resolve] with the active [NavigationContext] carried forward unchanged.
 */
@OptIn(ExperimentalComposeUiApi::class)
@Composable
fun RoleShellScaffold(
    context: NavigationContext,
    onSwitchPersona: (() -> Unit)? = null,
    /** A parsed but not yet authorized deep link / notification target (P0-8). */
    pendingDeepLink: NativeDeepLink? = null,
    sectionMemory: WorkspaceSectionMemory,
    onDeepLinkHandled: (() -> Unit)? = null,
    content: @Composable (destination: PrimaryDestination, context: NavigationContext) -> Unit
) {
    val navigation = remember(context.activeRole) { IANavigationContract.forRole(context.activeRole) }
    // Keyed on the whole context: a different wedding, assignment or scope is re-resolved from
    // the start rather than inheriting an authorization granted to another context.
    var authorization by remember(context) { mutableStateOf(RoleShellAuthorization.initial(context)) }
    // The tab that reads as selected is the last AUTHORIZED one — never an unresolved safe return.
    val selectedId = authorization.authorizedDestinationId

    val roleTag = context.activeRole.roleId

    // P0-8: the WHOLE deep link is resolved here — target wedding, destination, Level-2 section
    // and entity id — through the same gate as a tap. Nothing is discarded before authorization.
    LaunchedEffect(pendingDeepLink) {
        val link = pendingDeepLink ?: return@LaunchedEffect
        when (val resolution = DeepLinkRouter.resolve(link, context)) {
            is Entitlements.Resolution.Allowed -> {
                authorization = authorization.applying(resolution)
                // Level-2 deep links land on the requested section, not the workspace default.
                sectionMemory.applyRequested(
                    context = context,
                    destinationId = resolution.destination.id,
                    requested = (link as? NativeDeepLink.Workspace)?.section,
                    available = resolution.destination.sections
                )
            }
            is Entitlements.Resolution.Denied -> authorization = authorization.applying(resolution)
        }
        onDeepLinkHandled?.invoke()
    }

    // IA V2 §15.2 — tablets promote Level-1 into a navigation rail. The taxonomy is identical;
    // only the presentation adapts, so there is no separate tablet information architecture.
    BoxWithConstraints(modifier = Modifier.fillMaxSize()) {
        val useRail = maxWidth >= TABLET_RAIL_BREAKPOINT

        val onSelect: (PrimaryDestination) -> Unit = { destination ->
            authorization = authorization.selecting(context, destination.id)
        }

        // Content renders ONLY for a destination Entitlements.resolve allowed for this context.
        // Anything else is the access boundary, including when nothing was ever authorized.
        val workspace: @Composable () -> Unit = {
            val visible = authorization.visibleDestinationId?.let { navigation.destination(it) }
            if (visible != null) {
                content(visible, context)
            } else {
                AccessBoundaryNotice(
                    reason = authorization.denial?.reason
                        ?: "This workspace is not authorized for this session.",
                    onDismiss = if (authorization.canDismissDenial) {
                        { authorization = authorization.dismissingDenial() }
                    } else {
                        null
                    }
                )
            }
        }

        Scaffold(
            modifier = Modifier
                .semantics { testTagsAsResourceId = true }
                .testTag("role-shell-$roleTag")
                .testTag(if (useRail) "layout-rail" else "layout-bottom-bar"),
            containerColor = WeddingIdentityPalette.Ivory,
            topBar = {
                RoleContextBar(context = context, onSwitchPersona = onSwitchPersona)
            },
            bottomBar = {
                if (!useRail) {
                    Column(modifier = Modifier.background(WeddingIdentityPalette.IvorySoft)) {
                        HorizontalDivider(thickness = 1.dp, color = WeddingIdentityPalette.Hairline)
                        NavigationBar(
                            containerColor = WeddingIdentityPalette.IvorySoft,
                            tonalElevation = 0.dp
                        ) {
                            navigation.primary.forEach { destination ->
                                NavigationBarItem(
                                    modifier = Modifier.testTag("nav-$roleTag-${destination.id}"),
                                    selected = selectedId == destination.id,
                                    onClick = { onSelect(destination) },
                                    icon = {
                                        Icon(iconFor(destination.id), contentDescription = destination.label)
                                    },
                                    label = {
                                Text(
                                    destination.label,
                                    fontSize = 11.sp,
                                    maxLines = 1,
                                    // "Wedding Day" was being clipped to "Wedding". Scale the
                                    // label to fit instead of losing a word (accessibility +
                                    // Android/iOS label parity).
                                    softWrap = false,
                                    overflow = androidx.compose.ui.text.style.TextOverflow.Visible,
                                    style = androidx.compose.ui.text.TextStyle(
                                        fontSize = 11.sp,
                                        platformStyle = null
                                    ),
                                    modifier = Modifier.fillMaxWidth(),
                                    textAlign = androidx.compose.ui.text.style.TextAlign.Center
                                )
                            },
                                    colors = NavigationBarItemDefaults.colors(
                                        selectedIconColor = WeddingIdentityPalette.ChampagneDeep,
                                        selectedTextColor = WeddingIdentityPalette.ChampagneDeep,
                                        indicatorColor = Color.Transparent,
                                        unselectedIconColor = WeddingIdentityPalette.Muted,
                                        unselectedTextColor = WeddingIdentityPalette.Muted
                                    )
                                )
                            }
                        }
                    }
                }
            }
        ) { innerPadding ->
            Row(
                modifier = Modifier
                    .fillMaxSize()
                    .padding(innerPadding)
            ) {
                if (useRail) {
                    NavigationRail(
                        containerColor = WeddingIdentityPalette.IvorySoft,
                        // The rail must be given an explicit width. Inside a Row it otherwise
                        // accepts the whole proposal — measured at [0,1600] on a tablet — leaving
                        // the content's weight(1f) box zero width, so every tablet workspace
                        // rendered as an empty ivory page beside a stretched rail. The width is
                        // wide enough for "Wedding Day" at 11sp without truncating it.
                        modifier = Modifier
                            .width(TABLET_RAIL_WIDTH)
                            .fillMaxHeight()
                            .testTag("nav-rail-$roleTag")
                    ) {
                        navigation.primary.forEach { destination ->
                            NavigationRailItem(
                                modifier = Modifier.testTag("nav-$roleTag-${destination.id}"),
                                selected = selectedId == destination.id,
                                onClick = { onSelect(destination) },
                                icon = {
                                    Icon(iconFor(destination.id), contentDescription = destination.label)
                                },
                                label = {
                                Text(
                                    destination.label,
                                    fontSize = 11.sp,
                                    maxLines = 1,
                                    // "Wedding Day" was being clipped to "Wedding". Scale the
                                    // label to fit instead of losing a word (accessibility +
                                    // Android/iOS label parity).
                                    softWrap = false,
                                    overflow = androidx.compose.ui.text.style.TextOverflow.Visible,
                                    style = androidx.compose.ui.text.TextStyle(
                                        fontSize = 11.sp,
                                        platformStyle = null
                                    ),
                                    modifier = Modifier.fillMaxWidth(),
                                    textAlign = androidx.compose.ui.text.style.TextAlign.Center
                                )
                            },
                                colors = NavigationRailItemDefaults.colors(
                                    selectedIconColor = WeddingIdentityPalette.ChampagneDeep,
                                    selectedTextColor = WeddingIdentityPalette.ChampagneDeep,
                                    indicatorColor = Color.Transparent,
                                    unselectedIconColor = WeddingIdentityPalette.Muted,
                                    unselectedTextColor = WeddingIdentityPalette.Muted
                                )
                            )
                        }
                    }
                    VerticalDivider(thickness = 1.dp, color = WeddingIdentityPalette.Hairline)
                }

                Box(modifier = Modifier.weight(1f)) { workspace() }
            }
        }
    }
}

/** Width at which Level-1 moves from a bottom bar to a navigation rail (IA V2 §15.2). */
internal val TABLET_RAIL_BREAKPOINT = 600.dp

/** The rail's own width. Wide enough for the longest destination label without truncation. */
internal val TABLET_RAIL_WIDTH = 104.dp

/** IA V2 §1.2 — role and active wedding must never be ambiguous. */
@Composable
private fun RoleContextBar(
    context: NavigationContext,
    onSwitchPersona: (() -> Unit)?
) {
    Surface(color = WeddingIdentityPalette.IvorySoft) {
        Column {
            Row(
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(horizontal = 16.dp, vertical = 10.dp)
                    .testTag("role-context-bar"),
                verticalAlignment = Alignment.CenterVertically
            ) {
                Column(modifier = Modifier.weight(1f)) {
                    Text(
                        IANavigationContract.forRole(context.activeRole).displayName,
                        color = WeddingIdentityPalette.Ink,
                        fontFamily = FontFamily.Serif,
                        fontWeight = FontWeight.SemiBold,
                        fontSize = 17.sp,
                        maxLines = 1,
                        overflow = androidx.compose.ui.text.style.TextOverflow.Ellipsis
                    )
                    Text(
                        context.activeWeddingTitle,
                        color = WeddingIdentityPalette.Muted,
                        fontSize = 12.sp,
                        maxLines = 1,
                        // Without this the title is hard-clipped mid-word on a small phone —
                        // "Charity & Kudzie" became "Charity &", which reads as a rendering bug
                        // rather than as truncation.
                        overflow = androidx.compose.ui.text.style.TextOverflow.Ellipsis,
                        modifier = Modifier.testTag("active-wedding-label")
                    )
                }
                Surface(
                    shape = RoundedCornerShape(50),
                    color = WeddingIdentityPalette.ForestSoft
                ) {
                    Text(
                        context.environment.title,
                        color = WeddingIdentityPalette.Forest,
                        fontSize = 9.sp,
                        fontWeight = FontWeight.Bold,
                        modifier = Modifier
                            .padding(horizontal = 8.dp, vertical = 3.dp)
                            .testTag("environment-badge")
                    )
                }
                onSwitchPersona?.let {
                    IconButton(onClick = it, modifier = Modifier.testTag("role-switch-button")) {
                        Icon(
                            Icons.Default.ManageAccounts,
                            contentDescription = "Switch role",
                            tint = WeddingIdentityPalette.ChampagneDeep
                        )
                    }
                }
            }
            HorizontalDivider(thickness = 1.dp, color = WeddingIdentityPalette.Hairline)
        }
    }
}

/** IA V2 §14 — explain the boundary, leak no destination data, offer a safe return. */
@Composable
private fun AccessBoundaryNotice(reason: String, onDismiss: (() -> Unit)?) {
    Column(
        modifier = Modifier
            .fillMaxSize()
            .padding(24.dp)
            .testTag("access-boundary-notice"),
        verticalArrangement = Arrangement.Center,
        horizontalAlignment = Alignment.CenterHorizontally
    ) {
        Icon(
            Icons.Default.Lock,
            contentDescription = null,
            tint = WeddingIdentityPalette.Muted,
            modifier = Modifier.size(34.dp)
        )
        Spacer(modifier = Modifier.height(12.dp))
        Text(
            "Not available for this role",
            color = WeddingIdentityPalette.Ink,
            fontFamily = FontFamily.Serif,
            fontWeight = FontWeight.SemiBold,
            fontSize = 18.sp
        )
        Spacer(modifier = Modifier.height(6.dp))
        Text(
            reason,
            color = WeddingIdentityPalette.Muted,
            fontSize = 13.sp
        )
        Spacer(modifier = Modifier.height(16.dp))
        // Offered only when there is an authorized destination to go back to.
        if (onDismiss != null) {
            TextButton(onClick = onDismiss, modifier = Modifier.testTag("access-boundary-dismiss")) {
                Text("Go back", color = WeddingIdentityPalette.ChampagneDeep, fontWeight = FontWeight.SemiBold)
            }
        }
    }
}

/**
 * Level-2 workspace selector. Renders the documented section taxonomy as chips; depth lives here
 * rather than in the bottom bar (IA V2 §1.1).
 */
@Composable
fun WorkspaceSectionChips(
    sections: List<String>,
    selected: String,
    testTagPrefix: String,
    onSelect: (String) -> Unit
) {
    if (sections.isEmpty()) return
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .horizontalScroll(rememberScrollState())
            .padding(horizontal = 14.dp, vertical = 8.dp)
            .testTag("$testTagPrefix-sections"),
        horizontalArrangement = Arrangement.spacedBy(7.dp)
    ) {
        sections.forEach { section ->
            val isSelected = section == selected
            Surface(
                shape = RoundedCornerShape(50),
                color = if (isSelected) WeddingIdentityPalette.ChampagneDeep else WeddingIdentityPalette.IvorySoft,
                border = if (isSelected) null else androidx.compose.foundation.BorderStroke(1.dp, WeddingIdentityPalette.Hairline),
                onClick = { onSelect(section) },
                modifier = Modifier.testTag("$testTagPrefix-section-${section.slug()}")
            ) {
                Text(
                    section,
                    color = if (isSelected) WeddingIdentityPalette.IvorySoft else WeddingIdentityPalette.Muted,
                    fontSize = 11.sp,
                    fontWeight = if (isSelected) FontWeight.SemiBold else FontWeight.Normal,
                    maxLines = 1,
                    modifier = Modifier.padding(horizontal = 12.dp, vertical = 7.dp)
                )
            }
        }
    }
}

internal fun String.slug(): String =
    lowercase().replace(Regex("[^a-z0-9]+"), "-").trim('-')

private fun iconFor(destinationId: String): ImageVector = when (destinationId) {
    "home" -> Icons.Default.Home
    "plan", "workspace" -> Icons.Default.CalendarMonth
    "guests" -> Icons.Default.Group
    "wedding_day" -> Icons.Default.Celebration
    "clients" -> Icons.Default.FolderSpecial
    "daily_ops" -> Icons.Default.Bolt
    "invitation" -> Icons.Default.MailOutline
    "pass" -> Icons.Default.QrCode
    "jobs" -> Icons.Default.Storefront
    "schedule" -> Icons.Default.Schedule
    "messages" -> Icons.Default.Email
    "scan" -> Icons.Default.QrCodeScanner
    "admissions" -> Icons.Default.HowToReg
    "incidents" -> Icons.Default.ReportProblem
    "today" -> Icons.Default.Today
    "run_sheet" -> Icons.Default.Checklist
    "team" -> Icons.Default.Groups
    "dashboard" -> Icons.Default.Dns
    "cases" -> Icons.Default.SupportAgent
    "accounts" -> Icons.Default.ManageAccounts
    "audit" -> Icons.Default.Shield
    else -> Icons.Default.Menu
}

File: apps/ios/Wewed/Views/Roles/RoleShellScaffold.swift
import SwiftUI

/// Contract-driven Level-1 shell shared by every IA V2 role.
///
/// The tab bar is rendered from `IANavigationContract` rather than hand-written per role, so
/// labels and order cannot drift between roles or platforms. Every destination switch resolves
/// through `Entitlements.resolve` with the active `NavigationContext` carried forward unchanged.
public struct RoleShellScaffold<Content: View>: View {
    private let context: NavigationContext
    private let onSwitchPersona: (() -> Void)?
    /// A parsed but not yet authorized deep link / notification target (P0-8).
    private let pendingDeepLink: NativeDeepLink?
    private let onDeepLinkHandled: (() -> Void)?
    @ObservedObject private var sectionMemory: WorkspaceSectionMemory
    private let content: (PrimaryDestination, NavigationContext) -> Content

    /// Every render decision comes from here; see `RoleShellAuthorization`.
    @State private var authorization: RoleShellAuthorization

    /// The tab that reads as selected: the last AUTHORIZED one, never an unresolved safe return.
    private var selectedId: String {
        authorization.authorizedDestinationId ?? navigation.primary[0].id
    }

    #if os(iOS)
    @Environment(\.horizontalSizeClass) private var horizontalSizeClass
    #endif

    /// IA V2 §15.2 — regular width promotes Level-1 into a navigation rail. The taxonomy is
    /// identical; only the presentation adapts, so there is no separate tablet IA.
    private var useRail: Bool {
        #if os(iOS)
        return horizontalSizeClass == .regular
        #else
        return true
        #endif
    }

    public init(
        context: NavigationContext,
        onSwitchPersona: (() -> Void)? = nil,
        pendingDeepLink: NativeDeepLink? = nil,
        sectionMemory: WorkspaceSectionMemory,
        onDeepLinkHandled: (() -> Void)? = nil,
        @ViewBuilder content: @escaping (PrimaryDestination, NavigationContext) -> Content
    ) {
        self.context = context
        self.onSwitchPersona = onSwitchPersona
        self.pendingDeepLink = pendingDeepLink
        self.sectionMemory = sectionMemory
        self.onDeepLinkHandled = onDeepLinkHandled
        self.content = content
        // The initial destination is resolved through Entitlements like any other. It used to be
        // rendered directly, before any authorization had run (Phase 1 independent review).
        _authorization = State(initialValue: RoleShellAuthorization.initial(context))
    }

    private var navigation: RoleNavigation {
        IANavigationContract.forRole(context.activeRole)
    }

    public var body: some View {
        VStack(spacing: 0) {
            RoleContextBar(context: context, onSwitchPersona: onSwitchPersona)

            if useRail {
                HStack(spacing: 0) {
                    navigationRail
                    Divider().overlay(WeddingIdentityPalette.hairline)
                    workspace
                        .frame(maxWidth: .infinity, maxHeight: .infinity)
                }
                .accessibilityIdentifier("layout-rail")
            } else {
                TabView(selection: tabSelection) {
                    ForEach(navigation.primary) { destination in
                        // Every workspace renders inside the Wewed screen container, which
                        // publishes an explicit bounded width. Media then fills that width
                        // instead of dictating the screen's width (P0 responsive contract).
                        WewedScreenContainer {
                            // Only the destination Entitlements allowed renders content; every
                            // other tab page is the access boundary.
                            if authorization.visibleDestinationId == destination.id {
                                content(destination, context)
                            } else {
                                accessBoundary
                            }
                        }
                        .tabItem {
                            // The identifier must sit on the tab BUTTON, not on the tab's content
                            // view. Applied outside .tabItem it lands on the page, leaving the bar
                            // itself addressable only by its SF Symbol name — so a UI flow could
                            // not tap a tab at all, and "Guests" matched the Home metric tile
                            // instead. Android exposes the same nav-<role>-<destination> ids.
                            Label(destination.label, systemImage: Self.iconFor(destination.id))
                                .accessibilityIdentifier("nav-\(context.activeRole.roleId)-\(destination.id)")
                        }
                        .tag(destination.id)
                    }
                }
                .tint(WeddingIdentityPalette.champagneDeep)
                .accessibilityIdentifier("layout-bottom-bar")
            }
        }
        .background(WeddingIdentityPalette.ivory)
        .accessibilityIdentifier("role-shell-\(context.activeRole.roleId)")
        // P0-8: the WHOLE deep link is resolved here — target wedding, destination, Level-2
        // section and entity id — through the same gate as a tap. Nothing is discarded before
        // authorization.
        // A different wedding, assignment or scope is re-resolved from the start rather than
        // inheriting an authorization granted to another context (matches Android's keyed state).
        .onChange(of: context) { _, updated in
            authorization = RoleShellAuthorization.initial(updated)
        }
        .onChange(of: pendingDeepLink) { _, link in
            guard let link else { return }
            let resolution = DeepLinkRouter.resolve(link, context: context)
            authorization = authorization.applying(resolution)
            if case let .allowed(destination, _) = resolution {
                if case let .workspace(workspace) = link {
                    // Level-2 deep links land on the requested section, not the workspace default.
                    sectionMemory.applyRequested(
                        context: context,
                        destinationId: destination.id,
                        requested: workspace.section,
                        available: destination.sections
                    )
                }
            }
            onDeepLinkHandled?()
        }
    }

    /// Rail rendering of the very same contract destinations used by the tab bar.
    private var navigationRail: some View {
        VStack(spacing: 4) {
            ForEach(navigation.primary) { destination in
                Button {
                    tabSelection.wrappedValue = destination.id
                } label: {
                    VStack(spacing: 3) {
                        Image(systemName: Self.iconFor(destination.id))
                            .font(.system(size: 17))
                        Text(destination.label)
                            .font(.system(size: 11, weight: selectedId == destination.id ? .semibold : .regular))
                            .lineLimit(1)
                    }
                    .foregroundColor(
                        selectedId == destination.id
                            ? WeddingIdentityPalette.champagneDeep
                            : WeddingIdentityPalette.muted
                    )
                    // 48pt practical touch target (IA V2 §19).
                    .frame(width: 96, height: 56)
                    .contentShape(Rectangle())
                }
                .buttonStyle(.plain)
                .accessibilityIdentifier("nav-\(context.activeRole.roleId)-\(destination.id)")
            }
            Spacer()
        }
        .padding(.top, 10)
        .frame(width: 96)
        .background(WeddingIdentityPalette.ivorySoft)
        .accessibilityIdentifier("nav-rail-\(context.activeRole.roleId)")
    }

    @ViewBuilder
    private var workspace: some View {
        // Content renders ONLY for a destination Entitlements.resolve allowed for this context.
        if let visible = authorization.visibleDestinationId.flatMap(navigation.destination) {
            content(visible, context)
        } else {
            accessBoundary
        }
    }

    /// The boundary, with "Go back" offered only when there is an authorized destination to return
    /// to. Dismissing never reveals a destination that has not been resolved.
    private var accessBoundary: some View {
        AccessBoundaryNotice(
            reason: authorization.denialReason ?? "This workspace is not authorized for this session.",
            onDismiss: authorization.canDismissDenial
                ? { authorization = authorization.dismissingDenial() }
                : nil
        )
    }

    /// Every tab change passes through the entitlement gate before the destination is shown.
    private var tabSelection: Binding<String> {
        Binding(
            get: { selectedId },
            set: { requested in
                authorization = authorization.selecting(context, destinationId: requested)
            }
        )
    }

    static func iconFor(_ destinationId: String) -> String {
        switch destinationId {
        case "home": return "house.fill"
        case "plan", "workspace": return "calendar.badge.checkmark"
        case "guests": return "person.2.fill"
        case "wedding_day": return "sparkles"
        case "clients": return "folder.fill"
        case "daily_ops": return "bolt.fill"
        case "invitation": return "envelope.fill"
        case "pass": return "qrcode"
        case "jobs": return "bag.fill"
        case "schedule": return "clock.fill"
        case "messages": return "tray.full.fill"
        case "scan": return "qrcode.viewfinder"
        case "admissions": return "person.badge.check.mark.fill"
        case "incidents": return "exclamationmark.triangle.fill"
        case "today": return "sun.max.fill"
        case "run_sheet": return "list.bullet.rectangle.portrait"
        case "team": return "person.3.fill"
        case "dashboard": return "square.grid.2x2.fill"
        case "cases": return "lifepreserver.fill"
        case "accounts": return "person.crop.circle.fill"
        case "audit": return "lock.shield.fill"
        default: return "line.3.horizontal"
        }
    }
}

/// IA V2 §1.2 — role and active wedding must never be ambiguous.
struct RoleContextBar: View {
    let context: NavigationContext
    let onSwitchPersona: (() -> Void)?

    var body: some View {
        VStack(spacing: 0) {
            HStack(spacing: 10) {
                VStack(alignment: .leading, spacing: 1) {
                    Text(IANavigationContract.forRole(context.activeRole).displayName)
                        .font(.system(size: 17, weight: .semibold, design: .serif))
                        .foregroundColor(WeddingIdentityPalette.ink)
                        .lineLimit(1)
                    Text(context.activeWeddingTitle)
                        .font(.system(size: 12))
                        .foregroundColor(WeddingIdentityPalette.muted)
                        .lineLimit(1)
                        .accessibilityIdentifier("active-wedding-label")
                }

                Spacer()

                Text(context.environment.title)
                    .font(.system(size: 9, weight: .bold))
                    .padding(.horizontal, 8)
                    .padding(.vertical, 3)
                    .background(WeddingIdentityPalette.forestSoft)
                    .foregroundColor(WeddingIdentityPalette.forest)
                    .clipShape(Capsule())
                    .accessibilityIdentifier("environment-badge")

                if let onSwitchPersona {
                    Button(action: onSwitchPersona) {
                        Image(systemName: "person.crop.circle.badge.questionmark")
                            .foregroundColor(WeddingIdentityPalette.champagneDeep)
                    }
                    .accessibilityLabel("Switch role")
                    .accessibilityIdentifier("role-switch-button")
                }
            }
            .padding(.horizontal, 16)
            .padding(.vertical, 10)
            .accessibilityIdentifier("role-context-bar")

            Divider().overlay(WeddingIdentityPalette.hairline)
        }
        .background(WeddingIdentityPalette.ivorySoft)
    }
}

/// IA V2 §14 — explain the boundary, leak no destination data, offer a safe return.
struct AccessBoundaryNotice: View {
    let reason: String
    /// Nil when there is nowhere authorized to go back to.
    let onDismiss: (() -> Void)?

    var body: some View {
        VStack(spacing: 10) {
            Image(systemName: "lock.fill")
                .font(.system(size: 30))
                .foregroundColor(WeddingIdentityPalette.muted)
            Text("Not available for this role")
                .font(.system(size: 18, weight: .semibold, design: .serif))
                .foregroundColor(WeddingIdentityPalette.ink)
            Text(reason)
                .font(.system(size: 13))
                .foregroundColor(WeddingIdentityPalette.muted)
                .multilineTextAlignment(.center)
            if let onDismiss {
                Button("Go back", action: onDismiss)
                    .foregroundColor(WeddingIdentityPalette.champagneDeep)
                    .fontWeight(.semibold)
                    .accessibilityIdentifier("access-boundary-dismiss")
            }
        }
        .padding(24)
        .frame(maxWidth: .infinity, maxHeight: .infinity)
        .background(WeddingIdentityPalette.ivory)
        .accessibilityIdentifier("access-boundary-notice")
    }
}

/// Level-2 workspace selector. Renders the documented section taxonomy as chips; depth lives here
/// rather than in the tab bar (IA V2 §1.1).
public struct WorkspaceSectionChips: View {
    let sections: [String]
    let selected: String
    let testIdPrefix: String
    let onSelect: (String) -> Void

    public init(sections: [String], selected: String, testIdPrefix: String, onSelect: @escaping (String) -> Void) {
        self.sections = sections
        self.selected = selected
        self.testIdPrefix = testIdPrefix
        self.onSelect = onSelect
    }

    public var body: some View {
        if sections.isEmpty {
            EmptyView()
        } else {
            ScrollView(.horizontal, showsIndicators: false) {
                HStack(spacing: 7) {
                    ForEach(sections, id: \.self) { section in
                        let isSelected = section == selected
                        Button {
                            onSelect(section)
                        } label: {
                            Text(section)
                                .font(.system(size: 11, weight: isSelected ? .semibold : .regular))
                                .lineLimit(1)
                                .padding(.horizontal, 12)
                                .padding(.vertical, 7)
                                .background(isSelected ? WeddingIdentityPalette.champagneDeep : WeddingIdentityPalette.ivorySoft)
                                .foregroundColor(isSelected ? WeddingIdentityPalette.ivorySoft : WeddingIdentityPalette.muted)
                                .clipShape(Capsule())
                                .overlay(
                                    Capsule().stroke(
                                        isSelected ? Color.clear : WeddingIdentityPalette.hairline,
                                        lineWidth: 1
                                    )
                                )
                        }
                        .buttonStyle(.plain)
                        .accessibilityIdentifier("\(testIdPrefix)-section-\(section.iaSlug)")
                    }
                }
                .padding(.horizontal, 14)
                .padding(.vertical, 8)
            }
            .accessibilityIdentifier("\(testIdPrefix)-sections")
        }
    }
}

extension String {
    var iaSlug: String {
        lowercased()
            .replacingOccurrences(of: "[^a-z0-9]+", with: "-", options: .regularExpression)
            .trimmingCharacters(in: CharacterSet(charactersIn: "-"))
    }
}
